### PR TITLE
reimplement kube-addon-manager as a static Go binary

### PR DESCRIPTION
**Pull request by dekkagaijin**
This allows us to:
* implement unit tests
* eliminate dependencies on `bash` et al. and base the released image on `distroless`

I implemented this as a drop-in replacement for `addon-manager.sh`, but we can improve things over time. For example, we may want to use the k8s client directly and/or use watches to improve performance.

/kind cleanup
/kind feature

```release-note
NONE
```

### CODE REVIEW

**Comment by dekkagaijin** [comment]
/assign @mikedanese 
/assign @MrHohn 
PTAL

**Comment by MrHohn** [comment]
@dekkagaijin This is awesome :) Any high level description on what has changed in terms of functionalities? Do we expect this go program to behave ~ the same as the shell script version?

**Comment by dekkagaijin** [comment]
> Do we expect this go program to behave ~ the same as the shell script version?

Exactly. It should be a more-or-less drop in replacement for the bash script it replaces, but we can improve it over time.

**Comment by dekkagaijin** [comment]
gentle ping for any and all reviewers

**Comment by BenTheElder** [comment]
with 1.19 still struggling to come out of code freeze, more of us will have
time as 1.20 starts up officially.

**Comment by dekkagaijin** [comment]
@BenTheElder fair, just hoping to keep the ball rolling by getting the low-hanging-fruit comments

**Comment by zmerlynn** [comment]
/assign rngy

@rngy has volunteered to review it from our team (sorry, I let this sit and this week is no good for me, either).

File: cmd/kube-addon-manager/app/addon-manager.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+						time.Sleep(startAddonRetryDelay)
+					}
+				}
+			}
+			return err
+		})
+}
+
+// IsLeader returns whether this instance of addonManager should act as leader in a multi-master cluster.
+func (m *addonManager) isLeader() bool {
```
Comment (MrHohn) [inline cmd/kube-addon-manager/app/addon-manager.go:265]: I understand this is meant to keep the implementation at its current form. But has been wondering how easy would it be to implement a real leader election using https://github.com/kubernetes/client-go/blob/master/tools/leaderelection/leaderelection.go?

**Review by MrHohn — COMMENTED** [review]
Thanks for the works Jake, I left some comments for the first pass. This generally looks fine to me as the logic is mostly identical to the script.

File: cmd/kube-addon-manager/app/addon-manager.go (hunk 1 above)
Comment (dekkagaijin) [reply cmd/kube-addon-manager/app/addon-manager.go:265]: If it's not too bad I'll go ahead and do it, but want to get the direct port working first before implementing improvements

File: cmd/kube-addon-manager/README.md [hunk 2]
```diff
@@ -26,37 +26,3 @@ Otherwise it will be omitted.
 #### Images
 
 addon-manager images are pushed to `k8s.gcr.io`. As addon-manager is built for multiple architectures, there is an image per architecture in the format - `k8s.gcr.io/kube-addon-manager-$(ARCH):$(VERSION)`.
-
-#### How to release
```
[the remark this answers is not in the document]
Comment (dekkagaijin) [reply cmd/kube-addon-manager/README.md:30]: I was hoping to get some suggestions. There's a number of different methods leveraged across the different components, and haven't gotten a good answer as to the state-of-the-art in that regard

File: cmd/kube-addon-manager/addon-manager.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	"math/rand"
+	"os"
+	"time"
+
+	"k8s.io/component-base/logs"
+	"k8s.io/kubernetes/cmd/kube-addon-manager/app"
+)
+
+func main() {
+	rand.Seed(time.Now().UnixNano())
```
[the remark this answers is not in the document]
Comment (dekkagaijin) [reply cmd/kube-addon-manager/addon-manager.go:30]: Seemed like boilerplate in `main` across all of the components. It's a pretty common initialization step in languages which do not automatically seed their PRNGs

**Review by justaugustus — COMMENTED** [review]
@dekkagaijin -- As a meta question, given that this is a legacy directory, and our overall desire to decrease/remove the content from `cluster/`, does it make sense to maybe propose this as a separate repo?

I'm not sure who is ultimately responsible for addon-manager (minimal OWNERS in the leaves) or how it's released (the release instructions are no longer true given that we now use image promotion), but we can dig around together!

**Comment by dekkagaijin** [comment]
> @dekkagaijin -- As a meta question, given that this is a legacy directory, and our overall desire to decrease/remove the content from `cluster/`, does it make sense to maybe propose this as a separate repo?

It could? I don't know how those kind of decisions are made, but I put the new directory where I did to try and conform with _some_ existing pattern. I suppose it's a question of how substantial and independent the `addon-manager` is compared to the core k8s components?

> I'm not sure who is ultimately responsible for addon-manager (minimal OWNERS in the leaves) or how it's released (the release instructions are no longer true given that we now use image promotion), but we can dig around together!

Me neither, it seems to have been a very manual thing. I could add kube-addon-manager to the build/push logic used by the other components in that dir (apiserver, controller-manager)

**Comment by dims** [comment]
@cheftako do you have an opinion on the addon-manager?

File: cmd/kube-addon-manager/app/addon-manager.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+	m.waitForSystemServiceAccount()
+	if err := m.ensureDefaultAdmissionControlsObjects(); err != nil {
+		if os.IsNotExist(err) {
+			klog.Warningf("admissions control object directory not found: %v", err)
+		} else {
+			klog.Errorf("could not create admissions control objects: %v", err)
+		}
+	}
+	for {
+		start := time.Now()
```
[the remark this answers is not in the document]
Comment (cheftako) [reply cmd/kube-addon-manager/app/addon-manager.go:202]: +1

File: cmd/kube-addon-manager/app/addon-manager.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+			}
+			return err
+		})
+}
+
+// IsLeader returns whether this instance of addonManager should act as leader in a multi-master cluster.
+func (m *addonManager) isLeader() bool {
+	var out, errOut bytes.Buffer
+	klog.Info("== Determining addon-manager leader ==")
+	err := m.kubectl(&out, &errOut, m.kubectlOpts, "-n", systemNamespace, "get", "ep", "kube-controller-manager", "-o", `go-template={{index .metadata.annotations "control-plane.alpha.kubernetes.io/leader"}}`)
```
Comment (cheftako) [inline cmd/kube-addon-manager/app/addon-manager.go:268]: This is just gross. I realize this is legacy code and we need to make sure this code works correctly during an upgrade. (So we need to think through how to safely move off of the KCM lock) However piggy backing on the KCM's leader is *NOT* a good idea. 1) There are 3 possible resources for leader election, so we are not assured that its endpoints. 2) I believe the kube-up default is migrating away from endpoints and this will be broken when that migration completes. 3) The KCM leader is currently our heaviest burden on a control plane node. Deliberately co locating other process leaders with the KCM leader exacerbates this issue. 4) Leader election is timing sensitive with it durations. This code is beginning its belief it holds the lock value at some point after the lock was renewed and pays no attention to when the lock expires. As such it is not providing as good a guarantee to the addon manager as it does to the KCM.

Can we file as issue and provide a TODO in the code here related to the issue?

**Comment by cheftako** [comment]
> Walter Fender do you have an opinion on the addon-manager?

I've found it to be a fairly brittle part of our system. 
Switching the shell script for go binary seems like it is a start in fixing that.
Seems a bit odd having 1 go binary as a wrapper for another go binary but it still seems like a move in the right direction.

**Comment by dekkagaijin** [comment]
> Seems a bit odd having 1 go binary as a wrapper for another go binary but it still seems like a move in the right direction.

I agree, there's no reason why we shouldn't be able to rely on the APIs directly. For my own sanity and the sanity of reviewers, given that it wasn't tested previously, I tried to make this as direct a port as possible.

File: cmd/kube-addon-manager/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kube-addon-manager/README.md:30]: it should not be part of the release because the underlying mechanism was deprecated a long time ago.
the readme should reflect that.

new efforts around cluster addons are happening in kubernetes-sigs/cluster-addons

File: cmd/kube-addon-manager/addon-manager.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+	cmd := app.NewAddonManagerCommand()
+
+	logs.InitLogs()
+	defer logs.FlushLogs()
+
+	if err := cmd.Execute(); err != nil {
```
Comment (neolit123) [inline cmd/kube-addon-manager/addon-manager.go:36]: the main entrypoint should print a deprecation message.

**Review by neolit123 — COMMENTED** [review]
this feels like nudging the technical dept needle in the wrong direction.

> eliminate dependencies on bash et. al and base the released image on distroless

arguably the addon-manager image should not be released at this point.
the new tool under /cmd should not be released either.

> implement unit tests

unit tests are always a good argument, but not for deprecated mechanisms.

is there other agenda for executing on this action that was not mentioned in the OP?

a couple of my comments for approvers to consider:
https://github.com/kubernetes/kubernetes/pull/92978/files#r468526342

> it should not be part of the release because the underlying mechanism was deprecated a long time ago.
the readme should reflect that.
> new efforts around cluster addons are happening in https://github.com/kubernetes-sigs/cluster-addons

https://github.com/kubernetes/kubernetes/pull/92978/files#r468527148

> the main entrypoint should print a deprecation message, so that observers of the k/k code base do not use this in production. otherwise this new tool ends up promoting the deprecated tool under cluster/addons.

/cc @justinsb

**Comment by dekkagaijin** [comment]
@neolit123 The motivation is listed in the OP. It'd be great if the universe migrated to a better supported, more fully featured alternative, but until then we're going to have to maintain what we have right now.
In this case, the existing implementation of addon-manager relies on a "fat" OS image base that represents an ongoing maintenance headache (made unnecessarily risky by the dearth of tests). This PR lets us migrate to a minimal base.

**Comment by dims** [comment]
@dekkagaijin Jake, let's take this up in the next sig-arch call? Can you please talk to @johnbelamaric to get this on the agenda?

**Comment by dekkagaijin** [comment]
@dims the [regular sig meeting on Thursday](https://github.com/kubernetes/community/tree/master/sig-architecture#meetings)?

**Comment by dims** [comment]
this week friday for sure ... it's every other thu @dekkagaijin

**Comment by mikedanese** [comment]
I'm all for a long term solutions here (which is probably to delete cluster/, tell people to use any of N projects that do this better), but we haven't made great progress and I don't know of anyone actively working to finish it. 

Keeping up with debian base CVE patches has been very painful. I don't think we should block improvements that reduce churn and toil in the interim. Of course, inventing a new helm alternative in cluster/addons would be a bad outcome. Happy to discuss in sig-arch.

File: cmd/kube-addon-manager/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenTheElder) [reply cmd/kube-addon-manager/README.md:30]: I don't think cluster lifecycle can both disown a project and refuse changes to it ...

[the remark this answers is not in the document]
Comment (BenTheElder) [reply cmd/kube-addon-manager/README.md:30]: This shouldn't be part of the release build because we shouldn't need to re-build it on every commit / isn't tied to the kubernetes release lifecylce, as opposed to say, kube-scheduler, and it won't be bundled with the core release artifacts. Other images like this are the debian base images, go-runner, e2e images ...

There are other default images in the release build that are borderline inappropriate as-is (like the image consumed by sonobuoy), we should not extend that.

As for how you should release an image like this, we have automatic post-submit image builds used for e.g. the images under e2e, etc. a rabbit hole to go down:
https://github.com/kubernetes/k8s.io/tree/master/k8s.gcr.io
https://github.com/kubernetes/test-infra/tree/master/images/builder

Basically we automatically run a cloudbuild.yaml in post-submit, and then you can file a promotion from the staging registry to the prod registry in the k8s.io repo. Anything in the project can release there at any time.

**Comment by BenTheElder** [comment]
> this feels like nudging the technical dept needle in the wrong direction.

This moves us to having less bash running in CI etc. That is a good thing.

> is there other agenda for executing on this action that was not mentioned in the OP?

Yes ... see for example: `eliminate dependencies on bash et. al and base the released image on distroless`.

Moving to distroless is a strategic objective to secure the Kubernetes project and make things more maintainable.

> it should not be part of the release because the underlying mechanism was deprecated a long time ago.
the readme should reflect that.
new efforts around cluster addons are happening in https://github.com/kubernetes-sigs/cluster-addons

If it was actually deprecated that should have already happened. Failure to do so is part of a failure to properly deprecate it in the first place.

Deprecated or not this is still in use, and this does not actually add any features, only makes it easier to maintain.
Formal deprecation requires more than offhand stating that something is deprecated in a PR comment.

Forgive me for lack of better metrics but for cluster-addons, while I highly appreciate the people behind it and the effort...  64 stars / 18 watching / 27 forks / 13 contributors / < 200 commits over a year does not smell like a mature or high velocity project. Asking those vested in maintaining the addon manager to abandon it and make this new project successful in it's current state is a pretty big ask.

It does not appear to be a trivial migration, or something used in production.

The addon manager here is used in non-trivial production today.

I don't think it's a good look to disown a subproject ([cluster/ entirely](https://groups.google.com/a/kubernetes.io/d/msgid/steering/CAGDbWi-1Am_yQJ5qE5zx4%3DQsmpx1ndwy_9regCJApH6%3D3Mnx6w%40mail.gmail.com)) to be owned by other individuals / SIGs, and then circle back to block work to reduce maintenance concerns.

**Comment by BenTheElder** [comment]
A reminder that the bulk of CI depends on this and that it is a non-trivial, relatively unstaffed effort to change that.

I fully support getting this onto distroless through a straightforward port. 
I am also in favor of adding unit tests to prevent regressions.

File: cmd/kube-addon-manager/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kube-addon-manager/README.md:30]: > I don't think cluster lifecycle can both disown a project and refuse changes to it ...

it's not really a case where someone is "refusing" changes to this though.
all is recommendations.

that is why i've said:
> a couple of my comments for approvers to consider:

here:
https://github.com/kubernetes/kubernetes/pull/92978#pullrequestreview-465001212

**Comment by neolit123** [comment]
> I don't think it's a good look to disown a subproject (cluster/ entirely) to be owned by other individuals / SIGs, and then circle back to block work to reduce maintenance concerns.

that feels like a misinterpretation of my comments.

like i've said they are for approvers to consider:
https://github.com/kubernetes/kubernetes/pull/92978#pullrequestreview-465001212

...but it feels appropriate for deprecation notices to be added in any case.

**Comment by neolit123** [comment]
if supposedly this is only used in CI and is deprecated, why is the underlying image important WRT - CVEs vs distroless?
